format/ComputeIR.fbs

```diff
@@ -0,0 +1,210 @@
+  shape: Shape (required);
+
+  /// The type of this literal.
+  type: Type (required);
+
+  /// Buffers containing `length` elements of arrow-formatted data.
+  /// If `length` is absent (this Literal is scalar), these buffers
+  /// are sized to accommodate a single element of arrow-formatted data.
+  /// XXX this can be optimized for trivial scalars later
+  buffers: [Buffer];
+}
+
+table FieldRef {
+  /// A sequence of field names to allow referencing potentially nested fields
+  path: [string];
+
+  /// For Expressions which might reference fields in multiple Relations,
+  /// this index may be provided to indicate which Relation's fields
+  /// `path` points into. For example in the case of a join,
+  /// 0 refers to the left relation and 1 to the right relation.
+  relation_index: int;
+
+  /// The type of data in the referenced Field.
+  type: Type;
+}
+
+table Call {
+  /// The name of the function whose invocation this Call represents.
+  function_name: string (required);
+
+  /// Parameters for `function_name`; content/format may be unique to each
+  /// value of `function_name`.
+  options: Blob;
+
+  /// The arguments passed to `function_name`.
+  arguments: [Expression] (required);
+
+  /// The type of data which invoking `function_name` will return.
+  type: Type;
+}
+
+/// A relation is a set of rows with consitent schema.
+table Relation {
+  /// The namespaced name of this Relation.
+  ///
+  /// Names with no namespace are reserved for pure relational
+  /// algebraic operations, which currently include:
+  ///   "filter"
+  ///   "project"
+  ///   "aggregate"
+  ///   "join"
+  ///   "order_by"
+  ///   "limit"
+  ///   "literal"
+  ///   "interactive_output"
+  relation_name: string (required);
+
+  /// Parameters for `relation_name`; content/format may be unique to each
+  /// value of `relation_name`.
+  options: Blob;
+
+  /// The arguments passed to `relation_name`.
+  arguments: [Relation] (required);
+
+  /// The schema of rows in this Relation
+  schema: Schema;
+}
+
+/// The contents of Relation.options will be FilterOptions
+/// if Relation.name = "filter"
+table FilterOptions {
+  /// The expression which will be evaluated against input rows
+  /// to determine whether they should be excluded from the
+  /// "filter" relation's output.
+  filter_expression: Expression (required);
+}
+
+/// The contents of Relation.options will be ProjectOptions
+/// if Relation.name = "project"
+table ProjectOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "project" relation's output.
+  expressions: [Expression] (required);
+}
+
+/// The contents of Relation.options will be AggregateOptions
+/// if Relation.name = "aggregate"
+table AggregateOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "aggregate" relation's output.
+  aggregations: [Expression] (required);
+  /// Keys by which `aggregations` will be grouped.
+  keys: [Expression];
+}
```

Comment:
Do we need the ability to control `null` and/or `NaN` behavior here? Are they always grouped together?

Reply:
Again, I'd defer this configuration to a function call rather than introduce more configuration